stdlib/record.ncl

```diff
@@ -46,7 +46,66 @@
           false
         has_field "one" { one = 1, two = 2 } =>
           true
+      ```
       "%m
     = fun field r => %has_field% field r,
+
+    insert : forall a. Str -> a -> {_: a} -> {_: a}
+      | doc m%%"
+        Insert a new field in a record. `insert` doesn't mutate the original
+        record but return a new one instead.
+
+        ```nickel
+        insert "foo" foo { bar = "bar" } =>
+          { foo = "foo", bar = "bar }  
+
+        {}
+        |> insert "file.%{ext}" "data/text"
+        |> insert "length" 10*1000 =>
+          {"file.txt" = "data/text", "length" = 10000}
+        ```
+      "%%m
+      = fun field content r => %record_insert% field r content,
+
+    remove : forall a. Str -> {_: a} -> {_: a}
+      | doc m%"
+        Remove a field from a record. `remove` doesn't mutate the original
+        record but return a new one instead.
+
+        ```nickel
+        remove "foo" foo { foo = "foo", bar = "bar" } =>
+          { bar = "bar }  
+        ```
+      "%m
+      = fun field r => %record_remove% field r,
+
+    update | forall a. Str -> a -> { ; a} -> { ; a}
+      | doc m%"
+        Update a field of a record with a new value. `update` doesn't mutate the
+        original record but return a new one instead. If the field to update is absent
+        from the given record, `update` simply adds it.
+
+        ```nickel
+        remove "foo" foo { foo = "foo", bar = "bar" } =>
+          { bar = "bar" }  
+        ```
+
+        As opposed to overriding a value with the merge operator `&`, `update`
+        will only change the specified field and won't automatically update the other
+        fields which depend on it: 
+
+        ```nickel
+        { foo = bar + 1, bar | default = 0 } & { bar = 1 } =>
+          { foo = 2, bar = 1 }
+        update "bar" 1 {foo = bar + 1, bar | default = 0 } =>
+          { foo = 1, bar = 1 } 
+        ```
```

Comment:
So, Nickel being lazy, the expressions held in data structures are thunks. Thunks are suspended computations, and can be roughly considered to be equivalent to a function without parameter in a strict language. 

Then, we have `eval::Closure`. Usually, people uses the word closure to mean a function that captures its environment. In the Nickel codebase it is slightly more general, as it doesn't have to be a function, but can be any expression capturing an environment. A closure is thus just a term packed with its environment, mapping its free variables to other closures. In practice, it is really a tuple `(Term, Environment)` (well, a struct, to have named fields, but is isomorphic to that).

Then, how does it work for recursive records? Take `{foo = 1, bar = 1 + foo}`. After parsing, recursive records are standard hashmaps of expressions that happens to have free variables that can potentially refers to their siblings. At the first evaluation of a recursive records, the `eval::fixpoint::patch_field` function "ties the knot" by:

1. Building an environment with all the fields of the record inside (here `rec_env := foo -> (%1, env1), bar => (%2, env2)` where `%1` and `%2` are freshly allocated closures, and `env1` and `env2` are clones of the current environment of the record)
2. Going through the record fields and mutating their environment to include this `rec_env`: `env1 += rec_env, env2 += rec_env`.

If I understand your question, the I think the answer is `no`. Records are still hashmaps in the end, and `update` really just does an `insert`. The expression of the old field is not evaluated : 

```nickel
let failure = 1 | Str in # stupid failing contract
let r = record.update "foo" 1 { foo = failure, bar = foo } in
r.foo # Evaluates to 1 without issues
r.bar # Fail, because bar hasn't been updated and still refers to the original `foo` definition

let r2 = {foo | default = failure, bar = foo} & {foo = 1} in
r2.foo # ok, equals 1
r2.bar # by the magic of merging, equals to 1 too
```

Did I answer more or less to your question?

Reply:
Yes, this makes everything clearer, thanks a lot!  I was a bit confused about how records work and now I understand much better :-)

Reply:
I didn't think of them as capturing their own fields in their env like a closure, I think that's the piece I was missing.